### ISSUE TITLE
[JENKINS-71647] Add option to prevent git fetch for freestyle projects

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/git-plugin#3800]
<p>In our Jenkins build server we have a number of freestyle build jobs. Those build jobs are all working in the same workspace and each of those build jobs is calling another one after being finished: A->B->C->D->.... which is very time consuming. All build jobs have been configured for the same git repository to be able to track which developers contributed to the build and to checkout the same commit in each build. ... Some of those build jobs should be able to run in parallel since they only execute unit tests in separate subdirectories.</p>

<p>To minimize execution time of the whole build chain we wanted to execute build jobs B, C and D in parallel: A-><span class="error">&#91;B,C,D,...&#93;</span>.</p>

<p>Unfortunately, when B,C and D are started at the same time, they are all trying to fetch from the origin repository via git and store the update in the very same workspace's .git directory, which in the end results in failed builds for B, C and D.</p>

<p>The error message in the build looks similar to this in such cases:</p>

<p><tt>...</tt><br/>
<tt><span class="error">&#91;new branch&#93;</span> feature -> origin/feature</tt><br/>
<tt>error: cannot lock ref 'refs/remotes/origin/feature/task': is at d91b4e33b4a2d3404c0738c85b39314bb1d3c2be but expected 1b9026360df650a3ae1e224e09e92373fcafd9ef</tt></p>

<p>I think a workaround would be here to add an option in the git configuration to prevent polling of the git repository at all. Indeed, this option is already available for pipeline build jobs<img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/star_yellow.png" height="16" width="16" align="absmiddle" alt="" border="0"/>, but for freestyle ones it is missing.</p>

<p>Can we have this option implemented also for freestyle build jobs?</p>

<p><img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/star_yellow.png" height="16" width="16" align="absmiddle" alt="" border="0"/> <a href="https://www.jenkins.io/doc/pipeline/steps/git/" class="external-link" target="_blank" rel="nofollow noopener">https://www.jenkins.io/doc/pipeline/steps/git/</a> - Search for "Example: Git step with https protocol and polling disabled"</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=turboscholz">turboscholz</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-71647" target="_blank">Add option to prevent git fetch for freestyle projects</a></i></summary>
<i><ul>
<li><b>status</b>: Closed
<li><b>priority</b>: Minor
<li><b>component(s)</b>: git-plugin
<li><b>label(s)</b>: freestyle-job, git-plugin, plugin, polling
<li><b>resolution</b>: Won't Do
<li><b>resolved</b>: 2023-07-14T19:19:02+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 2
<li><b>imported</b>: 2025-12-02
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>In our Jenkins build server we have a number of freestyle build jobs. Those build jobs are all working in the same workspace and each of those build jobs is calling another one after being finished: A-&gt;B-&gt;C-&gt;D-&gt;.... which is very time consuming. All build jobs have been configured for the same git repository to be able to track which developers contributed to the build and to checkout the same commit in each build. ... Some of those build jobs should be able to run in parallel since they only execute unit tests in separate subdirectories.</p>

<p>To minimize execution time of the whole build chain we wanted to execute build jobs B, C and D in parallel: A-&gt;<span class="error">&#91;B,C,D,...&#93;</span>.</p>

<p>Unfortunately, when B,C and D are started at the same time, they are all trying to fetch from the origin repository via git and store the update in the very same workspace's .git directory, which in the end results in failed builds for B, C and D.</p>

<p>The error message in the build looks similar to this in such cases:</p>

<p><tt>...</tt>
<tt><span class="error">&#91;new branch&#93;</span> feature -&gt; origin/feature</tt>
<tt>error: cannot lock ref 'refs/remotes/origin/feature/task': is at d91b4e33b4a2d3404c0738c85b39314bb1d3c2be but expected 1b9026360df650a3ae1e224e09e92373fcafd9ef</tt></p>

<p>I think a workaround would be here to add an option in the git configuration to prevent polling of the git repository at all. Indeed, this option is already available for pipeline build jobs<img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/star_yellow.png" height="16" width="16" align="absmiddle" alt="" border="0"/>, but for freestyle ones it is missing.</p>

<p>Can we have this option implemented also for freestyle build jobs?</p>

<p><img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/star_yellow.png" height="16" width="16" align="absmiddle" alt="" border="0"/> <a href="https://www.jenkins.io/doc/pipeline/steps/git/" class="external-link" target="_blank" rel="nofollow noopener">https://www.jenkins.io/doc/pipeline/steps/git/</a> - Search for "Example: Git step with https protocol and polling disabled"</p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>Windows, current Git-plugin</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-71647?focusedId=439744&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-439744">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I don't plan to implement this feature.</p>

<p>The message "cannot lock ref" is usually an indication that more than one job is attempting to use the same directory for its Jenkins workspace.  If that is what is happening in your case, then you'll have unexpected results as the builds collide with each other in unexpected ways.  Let Jenkins determine the workspace so that it avoids the collisions for you.</p>
</pre>
</details>

</ul>
</details>

<p>I don't plan to implement this feature.</p>

<p>The message "cannot lock ref" is usually an indication that more than one job is attempting to use the same directory for its Jenkins workspace.  If that is what is happening in your case, then you'll have unexpected results as the builds collide with each other in unexpected ways.  Let Jenkins determine the workspace so that it avoids the collisions for you.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>turboscholz:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-71647?focusedId=439746&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-439746">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Yes, indeed the build jobs all share the same workspace because of a reason: Build job A is fetching the initial sources from git (after a build was triggered externally) and it builds the executables for the unit tests, which are executed in B, C, D and so on. Since we are working on a very big mono-repo it would be a huge waste of disk space to give each unit test build job an own workspace. (in our case the repo size is like &gt;2GB).</p>

<p>Apart from that, I wonder why the option to prevent polling is available for descriptive pipeline jobs, but not for freestyle jobs. Wouldn't it be good to align their options?</p>
</pre>
</details>

</ul>
</details>

<p>Yes, indeed the build jobs all share the same workspace because of a reason: Build job A is fetching the initial sources from git (after a build was triggered externally) and it builds the executables for the unit tests, which are executed in B, C, D and so on. Since we are working on a very big mono-repo it would be a huge waste of disk space to give each unit test build job an own workspace. (in our case the repo size is like >2GB).</p>

<p>Apart from that, I wonder why the option to prevent polling is available for descriptive pipeline jobs, but not for freestyle jobs. Wouldn't it be good to align their options?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-71647?focusedId=439750&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-439750">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I believe you will have undesirable surprises that are difficult to diagnose due to your choice to have multiple concurrent jobs share the same Jenkins workspace.</p>

<blockquote><p>I wonder why the option to prevent polling is available for descriptive pipeline jobs, but not for freestyle jobs. Wouldn't it be good to align their options?</p></blockquote>

<p>Pipeline polling is controlled at a level above the git plugin.</p>
</pre>
</details>

</ul>
</details>

<p>I believe you will have undesirable surprises that are difficult to diagnose due to your choice to have multiple concurrent jobs share the same Jenkins workspace.</p>

<blockquote><p>I wonder why the option to prevent polling is available for descriptive pipeline jobs, but not for freestyle jobs. Wouldn't it be good to align their options?</p></blockquote>

<p>Pipeline polling is controlled at a level above the git plugin.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-71647?focusedId=439751&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-439751">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>You might compare the cost of diagnosis with the cost of additional disc space.</p>

<p>Alternatively, you might consider improving the performance of workspace creation by using a reference repository that keeps a copy of the large monorepo on the agent and uses the reference repository feature to only copy content that is not already in the reference repository.  I created a <a href="https://youtu.be/jBGFjFc6Jf8?t=6434" class="external-link" target="_blank" rel="nofollow noopener">video</a> that describes reference repositories and other performance improvement tricks that I learned while wrestling with a 20+ GB repository.</p>
</pre>
</details>

</ul>
</details>

<p>You might compare the cost of diagnosis with the cost of additional disc space.</p>

<p>Alternatively, you might consider improving the performance of workspace creation by using a reference repository that keeps a copy of the large monorepo on the agent and uses the reference repository feature to only copy content that is not already in the reference repository.  I created a <a href="https://youtu.be/jBGFjFc6Jf8?t=6434" class="external-link" target="_blank" rel="nofollow noopener">video</a> that describes reference repositories and other performance improvement tricks that I learned while wrestling with a 20+ GB repository.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>turboscholz:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-71647?focusedId=439754&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-439754">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Thanks for the suggestion with reference repos. I stumbled over it yesterday already. It turned out that it would not be the ideal solution here since our Jenkins nodes are created dynamically in a cloud environment and their discs are also assigned dynamically. We would have to create a reference repository for every dynamically created Jenkins node by some scripts before the initial build (in build job A). This would be beyond the scope here.</p>

<p>Probably we will try to migrate our build jobs over to pipeline scripts then. At the moment I don't see a reason why the parallel unit test execution in subdirectories should influence each other. Do you foresee any possible problems I might not know? <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
</pre>
</details>

</ul>
</details>

<p>Thanks for the suggestion with reference repos. I stumbled over it yesterday already. It turned out that it would not be the ideal solution here since our Jenkins nodes are created dynamically in a cloud environment and their discs are also assigned dynamically. We would have to create a reference repository for every dynamically created Jenkins node by some scripts before the initial build (in build job A). This would be beyond the scope here.</p>

<p>Probably we will try to migrate our build jobs over to pipeline scripts then. At the moment I don't see a reason why the parallel unit test execution in subdirectories should influence each other. Do you foresee any possible problems I might not know? <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-71647?focusedId=439760&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-439760">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Since you provision new nodes dynamically, why not define the base operating system image of the new node to include the reference repository?  The reference repository becomes a standard part of the agent and each time the base operating system image is updated (new patches, new features, etc.), that gives an opportunity to refresh the reference repository.  That shifts a large portion of the 2 GB repository data transfer into the virtual machine cloning process (often using a local disc drive with deduplicated local file system) instead of performing a 2 GB repository data transfer over the network.</p>

<p>Including the reference repository will dramatically reduce the data transfer for initial clone.  That reduces the startup cost of the test environment and provides faster feedback to the people that want to know the results of the tests.</p>

<blockquote><p>I don't see a reason why the parallel unit test execution in subdirectories should influence each other. Do you foresee any possible problems I might not know?</p></blockquote>

<p>Since most of the developers probably don't run parallel tests in multiple subdirectories at the same time, they don't detect the assumptions they have made that cause tests to collide.  Whether those collisions are at the file system, the network, or some other resource, the parallel test environment will be checking test configurations that the developers have not checked.  When test environments check things, they tend to find things.  In this case, I think they will find things that the developers will probably declare to be irrelevant and artifacts of the test configuration, not a real world configuration.</p>
</pre>
</details>

</ul>
</details>

<p>Since you provision new nodes dynamically, why not define the base operating system image of the new node to include the reference repository?  The reference repository becomes a standard part of the agent and each time the base operating system image is updated (new patches, new features, etc.), that gives an opportunity to refresh the reference repository.  That shifts a large portion of the 2 GB repository data transfer into the virtual machine cloning process (often using a local disc drive with deduplicated local file system) instead of performing a 2 GB repository data transfer over the network.</p>

<p>Including the reference repository will dramatically reduce the data transfer for initial clone.  That reduces the startup cost of the test environment and provides faster feedback to the people that want to know the results of the tests.</p>

<blockquote><p>I don't see a reason why the parallel unit test execution in subdirectories should influence each other. Do you foresee any possible problems I might not know?</p></blockquote>

<p>Since most of the developers probably don't run parallel tests in multiple subdirectories at the same time, they don't detect the assumptions they have made that cause tests to collide.  Whether those collisions are at the file system, the network, or some other resource, the parallel test environment will be checking test configurations that the developers have not checked.  When test environments check things, they tend to find things.  In this case, I think they will find things that the developers will probably declare to be irrelevant and artifacts of the test configuration, not a real world configuration.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>turboscholz:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-71647?focusedId=439761&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-439761">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Oh, and one more thing: We are using the "<a href="https://plugins.jenkins.io/parameterized-trigger/" class="external-link" target="_blank" rel="nofollow noopener">Jenkins Parameterized Trigger plugin</a>" to "Pass-through Git Commit that was built" from one build job to another: Each build job in our chain does checkout the exact same commit from the fetched repo. Here I don't see why it should be necessary to poll the remove repository in every build job after job A. The git repository in the workspace should contain everything already, no poll should be needed. That's why I think a checkbox to prevent the polling would be a good thing. <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/lightbulb_on.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
</pre>
</details>

</ul>
</details>

<p>Oh, and one more thing: We are using the "<a href="https://plugins.jenkins.io/parameterized-trigger/" class="external-link" target="_blank" rel="nofollow noopener">Jenkins Parameterized Trigger plugin</a>" to "Pass-through Git Commit that was built" from one build job to another: Each build job in our chain does checkout the exact same commit from the fetched repo. Here I don't see why it should be necessary to poll the remove repository in every build job after job A. The git repository in the workspace should contain everything already, no poll should be needed. That's why I think a checkbox to prevent the polling would be a good thing. <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/lightbulb_on.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>turboscholz:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-71647?focusedId=439762&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-439762">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<blockquote><p>Since you provision new nodes dynamically, why not define the base operating system image of the new node to include the reference repository? <span class="error">&#91;...&#93;</span></p></blockquote>

<p>I will investigate this, thanks for the idea. <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/thumbs_up.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
</pre>
</details>

</ul>
</details>

<blockquote><p>Since you provision new nodes dynamically, why not define the base operating system image of the new node to include the reference repository? <span class="error">&#91;...&#93;</span></p></blockquote>

<p>I will investigate this, thanks for the idea. <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/thumbs_up.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-71647?focusedId=439767&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-439767">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Won't be implemented.</p>
</pre>
</details>

</ul>
</details>

<p>Won't be implemented.</p>
